narwhal/crypto/src/signature.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{CurrentNetwork, Digest, PrivateKey, PublicKey};
+
+use eyre::eyre;
+use rand::thread_rng;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{
+    mpsc::{channel, Sender},
+    oneshot,
+};
+
+pub type Signature = snarkvm_console::account::Signature<CurrentNetwork>;
+
+/// Wraps a private key and provides on-demand signing through a channel.
+// Note: this code is based on the fastcrypto implementation.
+#[derive(Clone)]
+pub struct SignatureService {
+    channel: Sender<(Digest, oneshot::Sender<Signature>)>,
+}
+
+impl SignatureService {
+    pub fn new(pk: PrivateKey) -> Self {
+        let (tx, mut rx): (Sender<(Digest, oneshot::Sender<Signature>)>, _) = channel(100);
+        tokio::spawn(async move {
+            while let Some((msg, sender)) = rx.recv().await {
+                // TODO(nkls): can we do better with the rng here?
+                let mut rng = thread_rng();
```

Comment:
yes, at least by assigning it before the loop

Reply:
ah but the rng is not `Send` 🤔

Reply:
Yeah, problem is `rx.recv().await`—we could lock it too but that's not awesome.